web/src/popover_menus.js

```diff
@@ -537,139 +537,160 @@ export function initialize() {
         },
     });
 
-    register_popover_menu("#stream_filters .topic-sidebar-menu-icon", {
-        ...left_sidebar_tippy_options,
-        onShow(instance) {
-            popover_instances.topics_menu = instance;
-            on_show_prep(instance);
-            const elt = $(instance.reference).closest(".topic-sidebar-menu-icon").expectOne()[0];
-            const $stream_li = $(elt).closest(".narrow-filter").expectOne();
-            const topic_name = $(elt).closest("li").expectOne().attr("data-topic-name");
-            const url = $(elt).closest("li").find(".topic-name").expectOne().prop("href");
-            const stream_id = stream_popover.elem_to_stream_id($stream_li);
-
-            instance.context = popover_menus_data.get_topic_popover_content_context({
-                stream_id,
-                topic_name,
-                url,
-            });
-            instance.setContent(parse_html(render_topic_sidebar_actions(instance.context)));
-        },
-        onMount(instance) {
-            const $popper = $(instance.popper);
-            const {stream_id, topic_name} = instance.context;
-
-            if (!stream_id) {
-                instance.hide();
-                return;
-            }
-
-            $popper.on("click", ".tab-option", (e) => {
-                $(".tab-option").removeClass("selected-tab");
-                $(e.currentTarget).addClass("selected-tab");
+    register_popover_menu(
+        "#stream_filters .topic-sidebar-menu-icon, .inbox-row .inbox-topic-menu",
+        {
+            ...left_sidebar_tippy_options,
+            onShow(instance) {
+                popover_instances.topics_menu = instance;
+                on_show_prep(instance);
+                let stream_id;
+                let topic_name;
+                let url;
+
+                if (instance.reference.classList.contains("inbox-topic-menu")) {
+                    const $elt = $(instance.reference);
+                    stream_id = Number.parseInt($elt.attr("data-stream-id"), 10);
+                    topic_name = $elt.attr("data-topic-name");
+                    url =
+                        window.location.protocol +
+                        "//" +
+                        window.location.host +
+                        "/" +
+                        $elt.attr("data-topic-url");
```

Comment:
We should continue to stay out of the business of construction URLs with `+`; e.g. using `new URL` or .

This particular place may be safe but unsafe URL construction is a common source for security bugs and we should make a habit of using the right APIs. I think `new URL(data_topic_url, page_params.realm_uri)` is roughly what we need?

Reply:
``` diff
diff --git a/web/src/popover_menus.js b/web/src/popover_menus.js
index d6d446937f..da6a2cc822 100644
--- a/web/src/popover_menus.js
+++ b/web/src/popover_menus.js
@@ -38,6 +38,7 @@ import * as message_lists from "./message_lists";
 import * as message_viewport from "./message_viewport";
 import * as narrow_state from "./narrow_state";
 import * as overlays from "./overlays";
+import {page_params} from "./page_params";
 import * as popover_menus_data from "./popover_menus_data";
 import * as popovers from "./popovers";
 import * as read_receipts from "./read_receipts";
@@ -552,12 +553,7 @@ export function initialize() {
                     const $elt = $(instance.reference);
                     stream_id = Number.parseInt($elt.attr("data-stream-id"), 10);
                     topic_name = $elt.attr("data-topic-name");
-                    url =
-                        window.location.protocol +
-                        "//" +
-                        window.location.host +
-                        "/" +
-                        $elt.attr("data-topic-url");
+                    url = new URL($elt.attr("data-topic-url"), page_params.realm_uri);
                 } else {
                     const $elt = $(instance.reference)
                         .closest(".topic-sidebar-menu-icon")
```

Reply:
Yeah, thanks!